Add source location to invalid_source_file_exceptiont

tautschnig opened the pull request:
This enables proper use in the C front-end in a way that is compatible
with the compiler-specific message handlers in goto-cc: those message
handlers generate source-location output that mimics that of the
compiler being emulated.

As future work it is thus possible to remove all "throw 0" instances
from the C front-end.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Add source location to invalid_source_file_exceptiont

This enables proper use in the C front-end in a way that is compatible
with the compiler-specific message handlers in goto-cc: those message
handlers generate source-location output that mimics that of the
compiler being emulated.

As future work it is thus possible to remove all "throw 0" instances
from the C front-end.

chrisr-diffblue commented on src/ansi-c/literals/convert_character_literal.cpp:
```diff
@@ -32,13 +33,13 @@ exprt convert_character_literal(
     std::basic_string<unsigned int> value=
       unescape_wide_string(std::string(src, 2, src.size()-3));
+    // the parser rejects empty character constants
+    CHECK_RETURN(!value.empty());
```
Nice :-)

chrisr-diffblue commented on src/ansi-c/literals/convert_character_literal.cpp:
```diff
@@ -92,9 +95,12 @@ exprt convert_character_literal(
+                                           source_location};
   }
 
+  result.add_source_location() = source_location;
```
Also a nice catch!

chrisr-diffblue commented on src/goto-programs/initialize_goto_model.cpp:
```diff
@@ -33,13 +33,12 @@ Author: Daniel Kroening, kroening@kroening.com
 {
-  auto const entry_function_sym =
-    goto_model.symbol_table.lookup(entry_function_name);
+  const irep_idt &entry_function_name = options.get_option("function");
```
Since this is a (slight) change in semantics for this function, such that it now _requires_ the `function` option to be set (previously the function name was passed in as an argument) - is it worth adding a CHECK_RETURN here?

tautschnig replied:
Good point! I've added the suggested `CHECK_RETURN`.

chrisr-diffblue commented on src/util/exception_utils.h:
```diff
@@ -164,17 +164,42 @@ class analysis_exceptiont : public cprover_exception_baset
 /// For example, if we get C source code that is not syntactically valid
 /// or that has type errors.
-class invalid_source_file_exceptiont : public cprover_exception_baset
+class invalid_source_file_exceptiont : public invalid_input_exceptiont
```
👍

chrisr-diffblue reviewed (approved):
That looks great to me, and is what we should have done when we first added the structured exceptions. Some nice code cleanups/refactoring along the way too.

chris-ryder reviewed (approved):
Previously reviewed and approved with my Diffblue account, re-approving with my non-Diffblue account to keep the CODEOWNER review.

martin-cs reviewed (approved):
Yes.  I don't think this breaks any scripting.

tautschnig committed:
Add tests covering error handling

None of our existing tests covers the failure paths of type checking
saturating arithmetic and character literal conversion, respectively.

peterschrammel commented on src/util/typecheck.cpp:
```diff
@@ -37,5 +38,11 @@ bool typecheckt::typecheck_main()
     error() << e << eom;
   }
 
+  catch(const invalid_source_file_exceptiont &e)
```
🎉
